File: web/src/types/tg-subscribe.ts
/** TG 影片订阅相关的类型，与后端 internal/tgsubscribe 的对外 DTO 一一对应。 */

export type TGMediaType = "movie" | "tv";
export type TGSubscriptionStatus = "active" | "paused" | "completed";
export type TGChannelStatus = "unknown" | "ok" | "error";
export type TGPushProvider = "auto" | "native" | "builtin";

/** 命中记录状态。unmatched/ambiguous/filtered/duplicate 是「没推」的四类原因。 */
export type TGRecordStatus =
  | "unmatched"
  | "ambiguous"
  | "filtered"
  | "duplicate"
  | "pending"
  | "pushed"
  | "upgraded"
  | "superseded"
  | "failed"
  | "ignored"
  /** 识别到了但投不出去：没有投递器、网盘不收这个协议，或账号没配转存凭据。不进重试。 */
  | "unsupported"
  /** 推送失败且**重试也不会好**（提取码错、分享失效、目录对不上）。同样不进重试。 */
  | "unretryable";

export interface TGChannel {
  id: number;
  chat_id: string;
  username: string;
  title: string;
  remark: string;
  level: number;
  enabled: boolean;
  status: TGChannelStatus;
  last_error: string;
  last_message_id: number;
  last_post_at?: string;
  matched_count: number;
  created_at?: string;
  /** 只在「新增频道」的响应里带值：首次回填落库了多少条历史帖。 */
  backfill_posts?: number;
  /**
   * 这个频道累计产出的匹配记录数。
   *
   * 与 matched_count 的区别很重要：matched_count 其实是「已见帖子数」。
   * 帖子数一直涨而产出一直是 0，就说明这个频道抓不到内容。
   */
  record_count: number;
}

/**
 * 内置推荐频道。
 *
 * 清单是实测筛出来的（见后端 internal/tgsubscribe/recommended_channels.go）：
 * 判据是「抽得出来 **而且投得出去**」——很多活跃频道发的全是夸克/百度分享链，
 * LitePan 抽不出来或投不出去，收录进来只会堆满「暂不支持投递」。
 *
 * 首次启动时它们已经**自动入库**成了普通频道，所以这里通常返回空（全被标成已添加）。
 * 还有内容只剩一种情况：当时没加成功（多半是还没配代理），用户后来配好了手动补加。
 * 用户自己删掉的那些不会重新出现 —— 播种是一次性的。
 */
export interface TGRecommendedChannel {
  username: string;
  /** 建议的备注名，入库时直接当备注用，之后想改就改。 */
  remark: string;
  /** 这个频道发什么、需要什么前提。 */
  summary: string;
  /** 当前用户的频道列表里是不是已经有了 —— 有了就不再单独列出来。 */
  added: boolean;
}

export interface TGTMDBSearchResult {
  id: number;
  media_type?: TGMediaType;
  title?: string;
  name?: string;
  original_title?: string;
  original_name?: string;
  overview?: string;
  poster_path?: string;
  release_date?: string;
  first_air_date?: string;
  vote_average?: number;
}

export interface TGSeasonInfo {
  season_number: number;
  episode_count: number;
  air_date?: string;
  name?: string;
}

export interface TGSubscription {
  id: number;
  tmdb_id: string;
  media_type: TGMediaType;
  title: string;
  original_title: string;
  year: number;
  poster_path: string;
  overview: string;
  aliases: string[] | null;
  alias_count: number;
  status: TGSubscriptionStatus;
  target_account_id: number;
  target_parent_id: string;
  target_display_path: string;
  quality_profile_id: number;
  push_provider: TGPushProvider;
  collect_window_min: number;
  upgrade_enabled: boolean;
  best_quality_score: number;
  matched_count: number;
  pushed_count: number;
  last_match_at?: string;
  last_push_at?: string;
  last_error: string;
  collected_episodes: number;
  aired_episodes: number;
  total_episodes: number;
  seasons?: TGSeasonInfo[];
  created_at?: string;
}

export interface TGSubscriptionEpisode {
  season: number;
  episode: number;
  record_id: number;
  offline_task_id: string;
  account_id: number;
  file_id: string;
  target_path: string;
  delivered_at?: string;
}

/**
 * 「已订阅」页的类型筛选。
 *
 * `jav`（番号）目前只是占位：后端 domain.TGMediaType 只有 movie / tv，
 * 订阅表里不会出现这个值 —— 选了它就是一个空列表加一句说明，
 * 而不是假装有数据。
 */
export type TGSubscribeKindFilter = "all" | "movie" | "tv" | "jav";

export const TG_KIND_FILTERS: { value: TGSubscribeKindFilter; label: string }[] = [
  { value: "all", label: "全部类型" },
  { value: "movie", label: "影片" },
  { value: "tv", label: "剧集" },
  { value: "jav", label: "番号" },
];

/**
 * 「已订阅」页的状态筛选。
 *
 * 取值刻意与订阅状态一一对应（`all` 之外都是 TGSubscriptionStatus 的字面量），
 * 这样状态下拉既当筛选器用、又直接是批量修改的目标值。
 */
export type TGSubscribeStatusFilter = "all" | TGSubscriptionStatus;

export const TG_STATUS_FILTERS: { value: TGSubscribeStatusFilter; label: string }[] = [
  { value: "all", label: "全部状态" },
  { value: "active", label: "订阅中" },
  { value: "paused", label: "暂停订阅" },
  { value: "completed", label: "完成订阅" },
];

/** 订阅状态的中文名，列表与详情弹窗共用同一份口径。 */
export function tgStatusLabel(status: TGSubscriptionStatus): string {
  switch (status) {
    case "active":
      return "订阅中";
    case "paused":
      return "暂停订阅";
    case "completed":
      return "完成订阅";
    default:
      return status;
  }
}


export interface TGQualityConfig {
  prefer_resolution: string[];
  prefer_codec: string[];
  prefer_source: string[];
  exclude_keywords: string[];
  min_resolution: string;
  max_size_gb: number;
  weights: Record<string, number>;
}

export interface TGQualityProfile {
  id: number;
  name: string;
  is_default: boolean;
  config: TGQualityConfig;
}

export interface TGMatchRecord {
  id: number;
  channel_id: number;
  chat_title: string;
  message_id: number;
  message_date?: string;
  raw_name: string;
  name_source: string;
  /** 资源类型：magnet / ed2k / share_115 / share_quark / http。 */
  resource_kind: string;
  /** 资源类型的中文名，直接显示用。 */
  kind_label: string;
  /** 资源的原始链接。字段名是历史遗留 —— ed2k 与分享链也存在这里，
   *  显示时请用 kind_label 而不是硬编码「磁力链接」。 */
  magnet: string;
  magnet_hash: string;
  size_bytes: number;
  parsed_title: string;
  parsed_year: number;
  /** -1 表示未识别 —— 0 是合法季号（特别篇），不能用 0 表达缺失。 */
  season: number;
  episode: number;
  episode_end: number;
  is_batch: boolean;
  resolution: string;
  video_codec: string;
  source_tag: string;
  subscription_id: number;
  subscription_title?: string;
  match_score: number;
  quality_score: number;
  status: TGRecordStatus;
  status_label: string;
  reason: string;
  offline_task_id: string;
  account_id: number;
  provider_kind: string;
  retry_count: number;
  next_retry_at?: string;
  created_at?: string;
}

/** TG 订阅的抓取配置。代理不在这里 —— 它是「系统设置 → 其他设置」里的全局项。 */
export interface TGConfig {
  enabled: boolean;
  auto_push: boolean;
  default_account_id: number;
  default_parent_id: string;
  default_display_path: string;
  default_quality_profile_id: number;
  collect_window_min: number;
  max_push_per_hour: number;
  /** 基准抓取间隔（秒）。实际间隔还会按频道优先级与频道数放大。 */
  poll_interval_sec: number;
  /** 新频道首次订阅时往回翻的页数，每页 20 条。0 表示只追新。 */
  backfill_pages: number;
  /** 实际生效的每频道间隔 —— 频道数 × 请求间隔 会抬高基准，这个值让用户看得见真实节奏。 */
  effective_interval_sec: number;
  channel_count: number;
  status: string;
  status_message: string;
  last_poll_at: string;
}

export interface TGConfigInput {
  enabled: boolean;
  auto_push: boolean;
  default_account_id: number;
  default_parent_id: string;
  default_display_path: string;
  default_quality_profile_id: number;
  collect_window_min: number;
  max_push_per_hour: number;
  poll_interval_sec: number;
  backfill_pages: number;
}

/** 体检报告里的一行：某种资源类型在最近一页帖子里命中了多少。 */
export interface TGProbeResourceCount {
  kind: string;
  /** 类型的中文名，直接显示。 */
  label: string;
  count: number;
  /**
   * 当前版本**在类型层面**能不能把它投递出去。
   * 「抓到了但投不了」要能和「压根抓不到」分开显示。
   *
   * 注意它是按频道算的，没有账号上下文：115 分享属于「类型支持、但要目标账号配了
   * 网页 Cookie 才真的能转存」，那种前提由 warnings 用文字交代，不在这里下结论。
   */
  deliverable: boolean;
}

/** 频道校验结果。网页预览模式不再有 bot 相关字段。 */
export interface TGChannelProbe {
  chat_id: string;
  username: string;
  title: string;
  type: string;
  post_count: number;
  latest_message_id: number;
  latest_post_at?: string;
  url_button_count: number;
  bare_button_count: number;
  /** 最近一页帖子里各类资源的命中情况（固定顺序，含命中 0 的类型）。 */
  resource_counts?: TGProbeResourceCount[];
  /** 正文链接里指向第三方站点的域名（最多 3 个）——非空且没抽到资源时，链接都在中转站上。 */
  external_hosts?: string[];
  /** 指向 @xxx_bot 的按钮/深链数，点进去私聊机器人才拿得到下载地址。 */
  bot_deep_links: number;
  /** 抽到了资源、但发布名不像影视资源因而不会落库的帖子数。 */
  discarded_posts: number;
  warnings?: string[];
}

/** 把体检的资源计数拼成一行可读文案，如「磁力 0 · ed2k 1 · 115 分享 15」。 */
export function formatProbeCounts(counts?: TGProbeResourceCount[]): string {
  if (!counts?.length) return "";
  return counts.map((c) => `${c.label} ${c.count}`).join(" · ");
}

/** 体检里「识别到了但投不出去」的那部分计数。 */
export function undeliverableProbeCounts(counts?: TGProbeResourceCount[]): TGProbeResourceCount[] {
  return (counts ?? []).filter((c) => c.count > 0 && !c.deliverable);
}

/** 一次历史搜索里某个频道的命中情况。 */
export interface TGChannelSearchHit {
  channel_id: number;
  name: string;
  matched: number;
}

/**
 * 历史搜索的结果。
 *
 * 「搜的是历史帖」这件事决定了它的产物：只落库、不推送，命中在匹配历史里
 * 以「待确认」出现，由用户确认后手动推。
 */
export interface TGHistorySearchResult {
  subscription_id: number;
  title: string;
  /** 实际用来搜的关键词（主标题 + 原名 + 别名，最多 3 个）。 */
  keywords: string[];
  channel_count: number;
  /** 计划发出的请求数（频道数 × 关键词数）—— 让用户对耗时与限流风险有预期。 */
  request_count: number;
  /** 实际失败的请求数（限流、超时等）。 */
  failed_requests: number;
  posts_scanned: number;
  /** 这次新落库的记录数。 */
  hit_records: number;
  channels?: TGChannelSearchHit[];
  /** 没有公开用户名、无法搜索的频道。 */
  skipped_channels?: string[];
  message: string;
}

export interface TGStats {
  status: {
    enabled: boolean;
    auto_push: boolean;
    connected: boolean;
    status: string;
    status_message: string;
    channel_count: number;
    subscription_count: number;
    poll_interval_sec: number;
    backfill_pages: number;
    last_poll_at: string;
  };
  records: Record<string, number>;
}

/** 试跑结果：把整条决策链摊开给用户看。 */
export interface TGQualityPreview {
  parsed: {
    raw: string;
    title_candidates: string[] | null;
    year: number | null;
    season: number | null;
    episode: number | null;
    episode_end: number | null;
    is_batch: boolean;
    resolution: string;
    video_codec: string;
    source: string;
    audio_codec: string;
    release_group: string;
    looks_like: boolean;
  };
  quality: { passed: boolean; score: number; reason: string };
  quality_profile_id: number;
  candidates: {
    subscription_id: number;
    title: string;
    year: number;
    media_type: TGMediaType;
    match_score: number;
    accepted: boolean;
    reason: string;
    title_hit: string;
    title_source: string;
  }[];
  unmatched_hint: string;
  thresholds: { accept: number; ambiguous: number };
}

export interface TGDiscoverPayload {
  page: number;
  total_pages: number;
  total_results: number;
  results: TGTMDBSearchResult[];
}

File: internal/domain/tg_subscribe.go
package domain

import (
	"context"
	"encoding/json"
	"time"
)

// 频道连通性状态。
const (
	TGChannelStatusUnknown = "unknown"
	TGChannelStatusOK      = "ok"
	TGChannelStatusError   = "error"
)

// 订阅生命周期状态。
const (
	TGSubStatusActive    = "active"
	TGSubStatusPaused    = "paused"
	TGSubStatusCompleted = "completed"
)

const (
	TGMediaTypeMovie = "movie"
	TGMediaTypeTV    = "tv"
)

// 命中记录状态。unmatched / ambiguous / filtered / duplicate 是「没推」的四类原因，
// 它们在匹配历史里要能被用户看见，所以必须落库而不是丢弃。
const (
	TGRecordUnmatched  = "unmatched"
	TGRecordAmbiguous  = "ambiguous"
	TGRecordFiltered   = "filtered"
	TGRecordDuplicate  = "duplicate"
	TGRecordPending    = "pending"
	TGRecordPushed     = "pushed"
	TGRecordUpgraded   = "upgraded"
	TGRecordSuperseded = "superseded"
	TGRecordFailed     = "failed"
	TGRecordIgnored    = "ignored"
	// TGRecordUnsupported 表示「识别到了，但当前投不出去」：
	// 要么资源类型没有投递器，要么目标网盘与内置下载器都不支持这个协议
	// （例如 ed2k 推给 123 账号），要么该账号没配转存需要的凭据。
	//
	// 与 failed 严格区分：failed 是「能投但这次失败了」，要退避重试；
	// unsupported 是确定性结论，**不进重试队列**，只作为可见性留在历史里。
	TGRecordUnsupported = "unsupported"
	// TGRecordUnretryable 表示「这次投放失败了，而且重试也不会好」。
	//
	// 典型的确定性失败：115 提取码错误、分享已失效/被删、目标目录对不上。
	// 与 failed 的区别只在**要不要重试**，与 unsupported 的区别在**原因性质**：
	// unsupported 是「这套组合天生投不了」，unretryable 是「这次的数据有问题」。
	//
	// 为什么要单开一个状态而不是复用 failed + retry_count 打满：
	// 后者会让界面显示「已重试 5 次」，而实际上一次都没重试过 —— 那是假的。
	// 更要紧的是，反复调 115 的 share/receive 正是触发风控的典型姿势
	// （官方 FAQ：「短时间内获取次数太多」）。
	TGRecordUnretryable = "unretryable"
)

// 推送通道。auto 表示按网盘离线下载能力自动选择。
const (
	TGPushProviderAuto    = "auto"
	TGPushProviderNative  = "native"
	TGPushProviderBuiltin = "builtin"
)

// TGChannel 是一个订阅源（Telegram 频道）。
//
// ChatID 存的是校验通过后固化的数字 id（如 -1001234567890）—— 公开频道的
// @username 在频道改名后会失效，只有数字 id 稳定，所以 Username 仅用于展示。
type TGChannel struct {
	ID            int64
	ChatID        string
	Username      string
	Title         string
	Remark        string
	Level         int
	Enabled       bool
	Status        string
	LastError     string
	LastMessageID int64
	LastPostAt    time.Time
	MatchedCount  int64
	CreatedAt     time.Time
	UpdatedAt     time.Time
}

// TGQualityConfig 是 tg_quality_profiles.config_json 的结构。
//
// 三个 prefer_* 是**有序列**：位置即优先级，第 0 名最高。
type TGQualityConfig struct {
	PreferResolution []string       `json:"prefer_resolution"`
	PreferCodec      []string       `json:"prefer_codec"`
	PreferSource     []string       `json:"prefer_source"`
	ExcludeKeywords  []string       `json:"exclude_keywords"`
	MinResolution    string         `json:"min_resolution"`
	MaxSizeGB        float64        `json:"max_size_gb"`
	Weights          map[string]int `json:"weights"`
}

type TGQualityProfile struct {
	ID        int64
	Name      string
	IsDefault bool
	Config    json.RawMessage
	CreatedAt time.Time
	UpdatedAt time.Time
}

// TGSubscription 是一条影片/剧集订阅。
//
// Aliases / Seasons 是创建订阅时从 TMDB 固化下来的快照 —— 匹配阶段只读这两份缓存，
// 不打 TMDB（见方案 D4）。Target* 三列为零值时表示「用全局默认」。
type TGSubscription struct {
	ID                int64
	TMDBID            string
	MediaType         string
	Title             string
	OriginalTitle     string
	Year              int
	PosterPath        string
	Overview          string
	Aliases           []string
	AliasesSyncedAt   time.Time
	Seasons           json.RawMessage
	SeasonsSyncedAt   time.Time
	Status            string
	TargetAccountID   int64
	TargetParentID    string
	TargetDisplayPath string
	QualityProfileID  int64
	PushProvider      string
	CollectWindowMin  int
	UpgradeEnabled    bool
	PendingDeadlineAt time.Time
	BestQualityScore  float64
	LastMatchAt       time.Time
	LastPushAt        time.Time
	MatchedCount      int64
	PushedCount       int64
	LastError         string
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

// TGSubscriptionEpisode 是一条已收集的剧集（电影固定 season/episode 为 0）。
type TGSubscriptionEpisode struct {
	ID             int64
	SubscriptionID int64
	Season         int
	Episode        int
	RecordID       int64
	OfflineTaskID  string
	AccountID      int64
	FileID         string
	TargetPath     string
	DeliveredAt    time.Time
}

// TGMatchRecord 是一条匹配记录，同时承担匹配历史、去重、失败重试三重职责。
//
// Season / Episode / EpisodeEnd 用 -1 表示「未识别」—— 0 是合法季号（特别篇），
// 所以不能用零值表达缺失。
type TGMatchRecord struct {
	ID          int64
	ChannelID   int64
	ChatTitle   string
	MessageID   int64
	MessageDate time.Time
	RawName     string
	NameSource  string
	// ResourceKind 是资源类型：magnet / ed2k / share_115 / share_quark / http。
	// 空值按 magnet 处理（半迁移状态与内存库测试）。
	ResourceKind string
	// Magnet 是资源的原始链接。字段名是历史遗留 —— ed2k 与分享链也存这里，
	// 见 telegram.ResourceRef.Raw。改名要动 DB 列，收益不抵风险。
	Magnet string
	// MagnetHash 是**资源指纹**，按 ResourceKind 加前缀，是跨类型去重键的一部分
	// （唯一索引 idx_tg_rec_msg_magnet / idx_tg_rec_sub_magnet）。
	// 各类型的格式见 telegram.ResourceRef.InfoHash 的注释。
	MagnetHash     string
	SizeBytes      int64
	ParsedTitle    string
	ParsedYear     int
	Season         int
	Episode        int
	EpisodeEnd     int
	IsBatch        bool
	Resolution     string
	VideoCodec     string
	SourceTag      string
	SubscriptionID int64
	MatchScore     float64
	QualityScore   float64
	Status         string
	Reason         string
	OfflineTaskID  string
	AccountID      int64
	TargetParentID string
	ProviderKind   string
	RetryCount     int
	NextRetryAt    time.Time
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// TGMatchRecordFilter 是匹配历史列表的筛选条件。
type TGMatchRecordFilter struct {
	ChannelID      int64
	SubscriptionID int64
	Status         string
	Keyword        string
	Limit          int
	Offset         int
}

type TGChannelRepository interface {
	Create(ctx context.Context, c *TGChannel) (int64, error)
	Update(ctx context.Context, c *TGChannel) error
	Delete(ctx context.Context, id int64) error
	Get(ctx context.Context, id int64) (*TGChannel, error)
	GetByChatID(ctx context.Context, chatID string) (*TGChannel, error)
	List(ctx context.Context, enabledOnly bool) ([]*TGChannel, error)
	MarkPost(ctx context.Context, id int64, messageID int64, at time.Time) error
	MarkStatus(ctx context.Context, id int64, status, lastError string) error
}

type TGQualityProfileRepository interface {
	Create(ctx context.Context, p *TGQualityProfile) (int64, error)
	Update(ctx context.Context, p *TGQualityProfile) error
	Delete(ctx context.Context, id int64) error
	Get(ctx context.Context, id int64) (*TGQualityProfile, error)
	GetDefault(ctx context.Context) (*TGQualityProfile, error)
	List(ctx context.Context) ([]*TGQualityProfile, error)
	SetDefault(ctx context.Context, id int64) error
}

type TGSubscriptionRepository interface {
	Create(ctx context.Context, s *TGSubscription) (int64, error)
	Update(ctx context.Context, s *TGSubscription) error
	Delete(ctx context.Context, id int64) error
	Get(ctx context.Context, id int64) (*TGSubscription, error)
	GetByTMDB(ctx context.Context, tmdbID, mediaType string) (*TGSubscription, error)
	List(ctx context.Context, status string) ([]*TGSubscription, error)
	// SetStatusBatch 一次改多条订阅的状态，返回实际改动的条数。
	//
	// 存在的理由是「已订阅」页的批量操作：逐条走 Update 需要先把每条完整读出来、
	// 再整行写回去 —— 那些请求里夹着别的字段，并发下会互相覆盖。
	// 这里只写 status 一列，读-改-写被压成一个 UPDATE。
	SetStatusBatch(ctx context.Context, ids []int64, status string) (int64, error)
	// ListPending 返回 pending_deadline_at 已到期的订阅，供 dispatcher 消费。
	ListPending(ctx context.Context, now time.Time) ([]*TGSubscription, error)
	// TouchPending 只在 deadline 为空或更早时前移 —— 窗口内的新候选不会无限延长窗口。
	TouchPending(ctx context.Context, id int64, deadline time.Time) error
	ClearPending(ctx context.Context, id int64) error
	MarkMatched(ctx context.Context, id int64, at time.Time) error
	MarkPushed(ctx context.Context, id int64, at time.Time, qualityScore float64) error
	MarkError(ctx context.Context, id int64, message string) error
}

type TGSubscriptionEpisodeRepository interface {
	Upsert(ctx context.Context, e *TGSubscriptionEpisode) error
	Has(ctx context.Context, subscriptionID int64, season, episode int) (bool, error)
	ListBySubscription(ctx context.Context, subscriptionID int64) ([]*TGSubscriptionEpisode, error)
	DeleteBySubscription(ctx context.Context, subscriptionID int64) (int64, error)
}

type TGMatchRecordRepository interface {
	// Create 用 ON CONFLICT DO NOTHING 插入；返回 id=0 表示这条记录已存在（重复）。
	Create(ctx context.Context, r *TGMatchRecord) (int64, error)
	Update(ctx context.Context, r *TGMatchRecord) error
	Get(ctx context.Context, id int64) (*TGMatchRecord, error)
	List(ctx context.Context, f TGMatchRecordFilter) ([]*TGMatchRecord, int, error)
	ListPendingBySubscription(ctx context.Context, subscriptionID int64) ([]*TGMatchRecord, error)
	ListRetryable(ctx context.Context, now time.Time, limit int) ([]*TGMatchRecord, error)
	// GetByOfflineTaskID 按离线任务 ID 反查记录（下载完成事件带回的就是这个 ID）。
	GetByOfflineTaskID(ctx context.Context, taskID string) (*TGMatchRecord, error)
	CountByStatus(ctx context.Context) (map[string]int, error)
	// CountByChannel 按频道统计匹配记录数。
	//
	// 用途是把「这个频道到底产出过东西没有」变成界面上看得见的一个数字：
	// 帖子数一直涨、产出一直是 0，用户自己就能判断这个频道抓不到内容
	// （最常见的原因是用「点击复制」按钮发资源，而网页预览不渲染这类按钮）。
	CountByChannel(ctx context.Context) (map[int64]int64, error)
	ClearBefore(ctx context.Context, before time.Time) (int64, error)
}
